### PR TITLE
CPM-137: duplicate a user (front)

### PR DESCRIPTION
**Pull request by nmarniesse**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/form/duplicate-secondary-action.tsx [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import {ReactView} from '@akeneo-pim-community/legacy-bridge/src/bridge/react';
+import {DuplicateOption} from './DuplicateOption';
+
+class DuplicateSecondaryAction extends ReactView {
+  reactElementToMount() {
+    return <DuplicateOption userId={this.getFormData().meta.id} userCode={this.getFormData().code} />;
```
Comment (nmarniesse) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/form/duplicate-secondary-action.tsx:7]: Maybe we can unify this file and `DuplicateOption.txs` but I don't really know how to do that

File: akeneo-design-system/src/components/Field/Field.tsx [hunk 2]
```diff
@@ -98,6 +106,11 @@ const Field = React.forwardRef<HTMLDivElement, FieldProps>(
           {incomplete && <IncompleteBadge />}
           <Label htmlFor={inputId} id={labelId}>
             {label}
+            {requiredLabel && (
```
Comment (pierallard) [inline akeneo-design-system/src/components/Field/Field.tsx:109]: You should add aria info
https://developer.mozilla.org/fr/docs/Accessibilit%C3%A9/ARIA/Techniques_ARIA/Utiliser_l_attribut_aria-required

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/datagrid/action/duplicate-user-action.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      ReactDOM.unmountComponentAtNode(container);
+      document.body.removeChild(container);
+    };
+    const onDuplicateSuccess = (duplicatedUserId: string) => {
+      closeApp();
+      Routing.redirect(Routing.generate('pim_user_edit', {identifier: duplicatedUserId}));
+    };
+
+    ReactDOM.render(
+      <>
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/datagrid/action/duplicate-user-action.tsx:26]: useless

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/form/DuplicateOption.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import {useToggleState} from '@akeneo-pim-community/shared';
+import {DuplicateUserApp} from '@akeneo-pim-community/user-ui';
+
+const Router = require('pim/router');
+const translate = require('oro/translator');
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/form/DuplicateOption.tsx:6]: import {useTranslate} from '@akeneo-pim-community/legacy-bridge';

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/form/duplicate-secondary-action.tsx (hunk 1 above)
Comment (pierallard) [reply src/Akeneo/UserManagement/Bundle/Resources/public/js/form/duplicate-secondary-action.tsx:7]: It's ok for me

File: src/Akeneo/UserManagement/Bundle/Resources/workspaces/user-ui/src/components/CreateUserForm.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    const response = await duplicateUser(router, userId, data);
+
+    if (null !== response && response.ok) {
+      response.json().then((data: any) => onSuccess(data.meta.id));
+    } else {
+      onError();
+      if (null !== response) {
+        response.json().then((data: any) => {
+          const newBackendErrors: BackendErrors = {};
+          data.values.forEach((errorValue: {path: string; message: string}) => {
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/workspaces/user-ui/src/components/CreateUserForm.tsx:41]: please check that data.values exists and is an array of {errorValue, message} before doing this line
(do nothing if it's not the case)

File: src/Akeneo/UserManagement/Bundle/Resources/workspaces/user-ui/tests/front/unit/components/CreateUserForm.unit.tsx [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      await userEvent.type(inputEmail, 'email');
+      fireEvent.submit(screen.getByTestId('form-create-user'));
+    });
+
+    expect(onError).toHaveBeenCalledTimes(1);
+    expect(await screen.findByText('passwords do not match')).toBeInTheDocument();
+    expect(await screen.findByText('email is not valid')).toBeInTheDocument();
+  });
+
+  test('it firs the cancel action', async () => {
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/workspaces/user-ui/tests/front/unit/components/CreateUserForm.unit.tsx:156]: it firs toi même

**Review by pierallard — APPROVED** [review]

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/form/DuplicateOption.tsx (hunk 4 above)
Comment (nmarniesse) [reply src/Akeneo/UserManagement/Bundle/Resources/public/js/form/DuplicateOption.tsx:6]: do you know why I can't do that?
image

Comment (pierallard) [reply src/Akeneo/UserManagement/Bundle/Resources/public/js/form/DuplicateOption.tsx:6]: Oh it has to be used in something having <DependenciesProvider> as parent (it's a context)

File: src/Akeneo/UserManagement/Bundle/Datagrid/Action/DuplicateUserAction.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+use Oro\Bundle\DataGridBundle\Extension\Action\Actions\NavigateAction;
+
+/**
+ * @author    Nicolas Marniesse <nicolas.marniesse@akeneo.com>
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+final class DuplicateUserAction extends NavigateAction
+{
+    protected $requiredOptions = [];
```
Comment (nmarniesse) [inline src/Akeneo/UserManagement/Bundle/Datagrid/Action/DuplicateUserAction.php:15]: private?

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/form/DuplicateOption.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+type DuplicateActionProps = {
```
Comment (nmarniesse) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/form/DuplicateOption.tsx:8]: DuplicateOptionProps

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/form/DuplicateOption.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  userId: number;
+  userCode: string;
+};
+
+const DuplicateOption = ({userId, userCode}: DuplicateActionProps) => {
```
Comment (nmarniesse) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/form/DuplicateOption.tsx:13]: DuplicateMenuLink

**Review by mmetayer — APPROVED** [review]

**Comment by nmarniesse** [comment]
to test: import/export

**Pull request merged**